Sources/MapboxCoreNavigation/EHorizon/RoadObjectDistanceInfo.swift

```diff
@@ -0,0 +1,134 @@
+     - parameter type: Road object type.
+     - parameter distanceToNearestEntry: Distance measured in meters to the nearest entry.
+     - parameter distanceToNearestExit: Distance measured in meters to nearest exit.
+     - parameter isInside: Boolean to indicate whether we're currently "inside" the object.
+     */
+    case polygon(identifier: RoadObjectIdentifier,
+                 type: RoadObjectType,
+                 distanceToNearestEntry: CLLocationDistance,
+                 distanceToNearestExit: CLLocationDistance,
+                 isInside: Bool)
+
+    /**
+     The information about distance to the road object represented as a subgraph.
+     - parameter identifier: Road object identifier.
+     - parameter type: Road object type.
+     - parameter distanceToNearestEntry: Distance measured in meters to the nearest entry.
+     - parameter distanceToNearestExit: Distance measured in meters to the nearest exit.
+     - parameter isInside: Boolean that indicates whether we're currently "inside" the object.
+     */
+    case subgraph(identifier: RoadObjectIdentifier,
+                  type: RoadObjectType,
+                  distanceToNearestEntry: CLLocationDistance,
+                  distanceToNearestExit: CLLocationDistance,
+                  isInside: Bool)
+    
+    /**
+     The information about distance to the road object represented as a line.
+     - parameter identifier: Road object identifier.
+     - parameter type: Road object type.
+     - parameter distanceToEntry: Distance from the current position to entry point measured in meters along the road graph. This value is 0 if already "within" the object.
+     - parameter distanceToExit" Distance from the current position to the most likely exit point measured in meters along the road graph.
+     - parameter distanceToEnd: Distance from the current position to the most distance exit point measured in meters along the road graph.
+     - parameter isEntryFromStart: Boolean that indicates whether we enter the road object from its start. This value is `false` if already "within" the object.
+     - parameter length: Length of the road object measured in meters.
+     */
+    case line(identifier: RoadObjectIdentifier,
+              type: RoadObjectType,
+              distanceToEntry: CLLocationDistance,
+              distanceToExit: CLLocationDistance,
+              distanceToEnd: CLLocationDistance,
+              isEntryFromStart: Bool,
+              length: CLLocationDistance)
+
+    /** Road object identifier */
+    public var roadObjectIdentifier: RoadObjectIdentifier {
+        switch self {
+        case .point(let identifier, _, _): fallthrough
+        case .gantry(let identifier, _, _): fallthrough
+        case .polygon(let identifier, _, _, _, _): fallthrough
+        case .subgraph(let identifier, _, _, _, _): fallthrough
+        case .line(let identifier, _, _, _, _, _, _):
+            return identifier
+        }
+    }
```

Comment:
I would move common fields back to how it was done with a the struct. But in case for this solution consider this style(and the same idea for the next property):
```suggestion
    public var roadObjectIdentifier: RoadObjectIdentifier {
        switch self {
        case .point(let identifier, _, _),
             .gantry(let identifier, _, _),
             .polygon(let identifier, _, _, _, _),
             .subgraph(let identifier, _, _, _, _),
             .line(let identifier, _, _, _, _, _, _):
            return identifier
        }
    }
```

Reply:
Flattening of `RoadObjectDistance` struct was done in part for usability purposes. Do you think it'll suffer this way?

Reply:
I mostly agree with @1ec5 comment below: https://github.com/mapbox/mapbox-navigation-ios/pull/2991#discussion_r632865124
The current implementation will require a user to either ignore with `_` or get these two properties in `switch` and `if case` every time they want to access distance info. And the case with `switch` might look like an overhead.

Reply:
As outlined in the mentioned comment branch, we are sticking with the flattened enum instead of an extra level of nesting. I think it may be easier to handle such structure in a notification. @1ec5 may add more context to this decision.